### ISSUE TITLE
foreign hostname setting shouldn't disable DNS completely

### ISSUE THREAD

**Issue by errordeveloper** [weaveworks/weave#1620]
Hostname currently takes precedence when weave derives the name to use in DNS. Having an option to ignore hostnames would be nice, but an additional fallback mode would be more helpful to the users.

At the moment a hostname value that doesn't end with `.weave.local` does take precedence, and I don't think it should, especially when the user has specified a label matcher and container has such label attached.

**Comment by rade** [comment]
> Having an option to ignore hostnames would be nice

And derive DNS names from containers/labels only? Do you have a compelling example for why that would be desirable?

> an additional fallback mode

like?

> At the moment a hostname value that doesn't end with `.weave.local` does effectively take precedence too, and I don't think it should, especially when the user has specified a label matcher and container has such label attached.

So again this is just "derive DNS names w/o paying attention to the hostname", correct?

There is an inherent relationship between hostnames and DNS names. Yes, there are no guarantees for that, but it is quite common. If we break it we need very good reasons to do so.

Sounds to me like you are fighting with some system here that has its own ideas for what hostnames should be. What purposes is it using them for?

**Comment by errordeveloper** [comment]
> There is an inherent relationship between hostnames and DNS names. Yes, there are no guarantees for that, but it is quite common. If we break it we need very good reasons to do so.

That makes sense in itself, however it's somewhat contradicted by the recommended mode where container name is used in DNS and hostname is a random string...

> Sounds to me like you are fighting with some system here that has its own ideas for what hostnames should be. What purposes is it using them for?

Yes, I've been trying to make Kubernetes work with weavedns as is.

If user could opt-in from checks for `.weave.local` and let it accept any short hostname (as long as it's set), it could work for Kubernetes. I also thought of `--hostname-only-from-label` and `--hostname-only-from-name`, as an explicit opt-in alternative to the automatic fallback I've originally imagined...

**Comment by errordeveloper** [comment]
I do still think that ideally a service instance should have a hostname and DNS A record with random suffix (`mysvc-<random_instance_id>.weave.local`) as well as group record (array of CNAME record) (`mysvc.weave.local`), which would fit nicer with Kubernetes model...

**Comment by rade** [comment]
> somewhat contradicted by the recommended mode where container name is used in DNS and hostname

Where does anybody recommend setting random hostnames?

> If user could opt-in from checks for `.weave.local` and let it accept any short hostname (as long as it's set), it could work for Kubernetes.

A container's domain is derived from the hostname. If we follow your proposal, then containers will be registered in DNS with a different domain (`weave.local`) than set in the container. This means looking up unqualified names won't work. (well, it does, but only thanks to the hack we put in place for alpine, which ignores domain settings). We could stick `weave.local` on the domain search path, but that immediately raises the issue of what to do when the user has already specified a path (cue yet more configuration options).

> I also thought of `--hostname-only-from-label` and `--hostname-only-from-name`, as an explicit opt-in alternative to the automatic fallback I've originally imagined

Thus stomping on any hostname already set?

**Comment by errordeveloper** [comment]
> Where does anybody recommend setting random hostnames?

My bad, I forgot weave overrides docker's hostnames...

> Thus stomping on any hostname already set?

True, it would be.

**Comment by 2opremio** [comment]
I am not sure how useful it is to use weaveDNS with Kubernetes at the moment. At the very least it can't be used as a replacement for the Kubernetes dns addon.

We could register podnames or containernames  as hostnames, but that would be incongruent with the dns addon of kubernetes, which registers service names.

I don't think that the user wants (or at least expects) podnames to be registered in DNS. Here's an example.

The default api server's service:

```
$ kubectl get svc kubernetes
NAME         LABELS                                    SELECTOR   IP(S)      PORT(S)
kubernetes   component=apiserver,provider=kubernetes   <none>     10.0.0.1   443/TCP
```

gets registered as `kubernetes.default.svc.cluster.local` with the dns addon (where `cluster.local` is configurable).

However, the corresponding pod's Docker container is called `k8s_apiserver.41647e53_k8s-master-127.0.0.1_default_a4e5633a912df483da39e118d2b4d230_9ba93174` where:

* `apiserver` is the container name
* `master-127.0.0.1` is the pod name
* `default` is the namespace

I guess we could register something like `apiserver.master-127.0.0.1.default.containers.cluster.local` but:

* I am not even sure weaveDNS supports registering subdomains instead of hostnames
* In case there were multiple pods, you would register the container in all pods to the same name (which may not be what the user wants).

**Comment by errordeveloper** [comment]
> I don't think that the user wants (or at least expects) podnames to be registered in DNS. Here's an example.

Yes, and the hostnames are actually podnames and not service names.

**Comment by 2opremio** [comment]
Not as per the example above. kubernetes is a service name not a pod name.

**Comment by rade** [comment]
This issue is meandering a bit.

What are we trying to achieve here?

**Comment by 2opremio** [comment]
> This issue is meandering a bit.

Yep, sorry for derailing it, I'll leave to @errordeveloper now.

**Comment by bboreham** [comment]
@errordeveloper can you summarise what this issue is about; what change we need to make to weave?
